documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc

```diff
@@ -0,0 +1,232 @@
+* `-XX:+ UseConcMarkSweepGC` - Enables usage of the CMS collector.
+* `-XX:+ CMSClassUnloadingEnabled` - Allows class unloading when the CMS collector is enabled.
+* `-XX:+ UseParNewGC` - Utilize a parallel collector for the young generation. This parameter minimizes pausing by using multiple collection threads in parallel.
+* `-XX:+ DisableExplicitGC` - Prevent explicit garbage collections.
+
+=== Other settings
+
+There are two additional parameters which are suggested to be used:
+
+* `-server` - Enables server mode for the JVM.
+* `-XX:+ UseLargePages` - Instructs the JVM to allocate memory in Large Pages. These pages must be configured at the OS level for this parameter to function successfully.
+
+=== Example configuration
+
+.32GB JVM
+----
+-server
+-Xmx32G
+-Xms32G
+-Xmn8G
+-XX:+UseLargePages
+-XX:+UseConcMarkSweepGC
+-XX:+UseParNewGC
+-XX:+DisableExplicitGC
+----
+
+== Network configuration
+
+Infinispan uses TCP/IP for sending packets over the network.
+
+In order to achieve the best results, it is recommended to increase TCP send and receive window size (refer to you OS manual for instructions).
+The recommended values are:
+
+* send window size - 640 KB
+* receive window size - 25 MB
+
+== Number of threads (Server mode only)
+
+Hot Rod Server uses worker threads which are activated by a client's requests.
+It's important to match the number of worker threads to the number of concurrent client requests:
+
+[source,java]
+.Hot Rod Server worker thread pool size
+----
+<hotrod-connector socket-binding="hotrod" cache-container="local" worker-threads="200">
+   <!-- Additional configuration here -->
+</hotrod-connector>
+----
+
+== Cache Store performance
+
+In order to achieve the best performance, please follow the recommendations below when using Cache Stores:
+
+* Use async mode (write-behind) if possible
+* Prevent cache misses by preloading data
+* For JDBC Cache Store:
+** Use indexes on `id` column to prevent table scans
+** Use PRIMARY_KEY on `id` column
+** Configure batch-size, fetch-size, etc
+
+== Hints for program developers
+
+There are also several hints for developers which can be easily applied to the client application and will boost up the performance.
+
+=== Ignore return values
+
+When you're not interested in returning value by `#put` or `#remove` method, use `Flag.IGNORE_RETURN_VALUES` flag as shown below:
+
+[source,java]
+.Flag.IGNORE_RETURN_VALUES
+----
+cache.getAdvancedCache().withFlags(Flag.IGNORE_RETURN_VALUES)
+----
+
+=== Skip cache load
+
+When you don't want to load an entry from a Cache Store, use `Flag.SKIP_CACHE_LOAD` as shown below:
+
+[source,java]
+.Flag.SKIP_CACHE_LOAD
+----
+cache.getAdvancedCache().withFlags(Flag.SKIP_CACHE_LOAD)
+----
+
+=== Use `Externalizable` for marshalling
+
+Infinispan uses JBoss Marshalling to transfer objects over the wire. The most efficient way to marshall user data is to provide an Externalizer:
```

Comment:
That's a lot of very useful information :) I'll try to summarize that and hide some of the details you mentioned (I wouldn't like to mention too much of our internal behavior since it can get out of date very quickly)